returnn/tf/layers/basic.py

```diff
@@ -2685,6 +2685,83 @@ def get_out_data_from_opts(cls, name, axis, dim=1, sources=(), **kwargs):
     return data
 
 
+class RepeatLayer(_ConcatInputLayer):
+  """
+  A wrapper around tf.repeat, but supports an additional batch axis for the durations
+  The sum of the repetitions has to be non-zero for each sequence in the batch.
+
+  This layer can only be used with Tensorflow 1.15.0 or newer.
+  """
+  layer_class = "repeat"
+
+  def __init__(self, repetitions, axis="T", **kwargs):
+    """
+    :param LayerBase repetitions: number of repetitions for each sequence and position in target axis, [B,T] or [T,B]
+    :param str|int axis: (dynamic) axis for repetition (currently only time axis is supported)
+    """
+    super(RepeatLayer, self).__init__(**kwargs)
+    repetitions_data = repetitions.output
+    input_data = self.input_data
+    assert self.input_data
+    original_axis = input_data.get_axis_from_description(axis)
+    if original_axis < input_data.batch_dim_axis:
+      input_data = input_data.copy_as_batch_major().copy_move_axis(original_axis+1, 1)
+    else:
+      input_data = input_data.copy_as_batch_major().copy_move_axis(original_axis, 1)
+    repetitions_data = repetitions_data.copy_as_batch_major()
+    # pad the target axis
+    paddings = [(0, 1) if i == 1 else (0, 0) for i in range(input_data.batch_ndim)]
+    padded_data = tf.pad(input_data.placeholder, paddings)  # [B, T+1, ...]
+    # those are the sequence lengths after expansion
+    target_seq_len = tf.reduce_sum(repetitions_data.placeholder, axis=1)  # [B], the new size_placeholder for 'T'
+    # maximum sequence length after expansion
+    max_duration = tf.reduce_max(target_seq_len)  # [] == T'
+    # the new padding is the difference of the maximum to the new target
+    target_padding_steps = tf.expand_dims(max_duration - target_seq_len, 1)  # [B, 1]
+    # add the repetitions for the artificial padding position
+    target_repetitions = tf.concat([repetitions_data.placeholder, target_padding_steps], axis=1)  # [B, T+1]
+    # flatten batch and time
+    shape = tf.shape(padded_data)
+    flat_shape = tf.concat([[shape[0] * shape[1]], shape[2:]], axis=0)
+    reshaped_data = tf.reshape(padded_data, flat_shape)  # [B * T+1, F]
+    reshaped_repetitions = tf.reshape(target_repetitions, (shape[0] * shape[1],))  # [B * T+1]
+    # run the repetition
+    repeated_data = tf.repeat(reshaped_data, reshaped_repetitions, axis=0)  # [B * T', ...]
+    # unflatten the output
+    target_shape = tf.concat([[shape[0], max_duration], shape[2:]], axis=0)
+    self.output.placeholder = tf.reshape(repeated_data, target_shape)  # [B, T', ...]
+    # set the size_placeholder
+    self.output.size_placeholder = self.input_data.size_placeholder
+    self.output.size_placeholder[0] = target_seq_len
```

Comment:
This is wrong. You are overwriting the input size placeholder here.
Also, again the code is too complicated.
This is simpler and fixes the problem:
```suggestion
    self.output.size_placeholder = {0: target_seq_len}
```

Reply:
I thought when I copy a data object, I will get a "new" instance of the `size_placeholder` dictionary. You code is not correct when there are multiple entries in the `size_placeholder`, which should be supported (but is untested) by the RepeatLayer

Reply:
But `self.input_data` is not necessarily a new instance.
Also, why do you want to modify `self.input_data` here? It's just very non-intuitive. Leave `input_data` as-is. You never should modify the `size_placeholder` of another `Data` (except the `Data` you currently create here, i.e. `self.output`). This could easily lead to bugs in other contexts.

Reply:
Also, why `self.input_data`, and not `input_data`. You potentially transposed the axes in some way, so `self.input_data.size_placeholder` might be wrong.

Reply:
Ah, of course my code is wrong as well, when the spatial axis was at a different position at the beginning.

Reply:
Your replies were not yet loaded for some reason. So I assume I should always create a new dictionary, I can do that.